File: rust/src/day20.rs
pub struct Submarine {
    enhacement_algorithm: Vec<bool>,
    input_image: Vec<Vec<bool>>,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            enhacement_algorithm: Vec::new(),
            input_image: Vec::new(),
        }
    }

    pub fn on_input(&mut self, result: &str) {
        if self.enhacement_algorithm.len() == 0 {
            self.enhacement_algorithm = result.chars().map(|c| c == '#').collect();
            return;
        }

        if result.is_empty() {
            return;
        }

        let mut curr_line: Vec<bool> = vec![false, false].into_iter().chain(result.chars().map(|c| c == '#')).collect();
        curr_line.extend([false, false]);

        if self.input_image.len() == 0 {
            self.input_image.push(vec![false; curr_line.len()]);
            self.input_image.push(vec![false; curr_line.len()]);
        }

        self.input_image.push(curr_line);
    }

    fn finish_padding(&mut self) {
        self.input_image.push(vec![false; self.input_image[0].len()]);
        self.input_image.push(vec![false; self.input_image[0].len()]);
    }

    fn extend_padding(&mut self) {
        let padding_value = self.input_image[0][0];

        self.input_image.iter_mut().for_each(|row| {
            row.insert(0, padding_value);
            row.push(padding_value);
        });

        let width = self.input_image[0].len();

        self.input_image.insert(0, vec![padding_value; width]);
        self.input_image.push(vec![padding_value; width]);
    }

    fn enhance(&mut self) {
        // We skip the outer layer of padding because it will always have the same value, but we need to go through the inner layer of padding. If it changes, we must expand the padding.
        let height = self.input_image.len() - 2;
        let width = self.input_image[0].len() - 2;

        // Figuring out the padding value.
        let padding_value = if self.input_image[0][0] {
            self.enhacement_algorithm[511]
        } else {
            self.enhacement_algorithm[0]
        };

        let mut result = vec![vec![padding_value; width + 2]; height + 2];
        let mut padding_changed = false;

        for y in 1..=height {
            for x in 1..=width {
                let num =
                    (self.input_image[y - 1][x - 1] as usize) << 8 |
                    (self.input_image[y - 1][x] as usize) << 7 |
                    (self.input_image[y - 1][x + 1] as usize) << 6 |
                    (self.input_image[y][x - 1] as usize) << 5 |
                    (self.input_image[y][x] as usize) << 4 |
                    (self.input_image[y][x + 1] as usize) << 3 |
                    (self.input_image[y + 1][x - 1] as usize) << 2 |
                    (self.input_image[y + 1][x] as usize) << 1 |
                    (self.input_image[y + 1][x + 1] as usize);

                result[y][x] = self.enhacement_algorithm[num];

                // If padding changed value, mark the flag so we can expand the padding.
                if x == 1 || x == width || y == 1 || y == height {
                    padding_changed = padding_changed || result[y][x] != padding_value;
                }
            }
        }

        self.input_image = result;

        if padding_changed {
            self.extend_padding();
        }
    }

    fn lit_pixels(&self) -> u32 {
        self.input_image.iter().map(|row| row.iter().fold(0, |s, p| s + (*p as u32))).sum()
    }

    fn _print_image(&self) {
        self.input_image.iter().for_each(|row| {
            row.iter().for_each(|p| {
                print!("{}", if *p { "#" } else { "." });
            });
            println!()
        });
    }

    pub fn output(&mut self) {
        self.finish_padding();

        for _ in 0..2 {
            self.enhance();
        }

        println!("Part 1: {}", self.lit_pixels());

        for _ in 0..48 {
            self.enhance();
        }

        println!("Part 2: {}", self.lit_pixels());
    }
}

File: 2021/julia/Day20.jl
module Day20

export Submarine, on_input!, output

mutable struct Submarine
    enhancement_algorithm::BitVector
    input_image::BitMatrix

    Submarine() = new(BitVector(undef, 0), BitMatrix(undef, 0, 0))
end

function on_input!(s::Submarine, a::AbstractString)
    curr_line = BitVector(map(x -> x == "#" ? 1 : 0, split(a, "")))

    if length(s.enhancement_algorithm) == 0
        s.enhancement_algorithm = curr_line
        return
    end

    isempty(a) && return

    prepend!(curr_line, [0, 0])
    append!(curr_line, [0, 0])

    if size(s.input_image) == (0, 0)
        # Padding on top.
        s.input_image = falses(length(curr_line), 2)
    end

    s.input_image = [s.input_image curr_line]
end

function finish_padding!(s::Submarine)
    width, _ = size(s.input_image)

    s.input_image = [s.input_image falses(width, 2)]
end

function extend_padding!(s::Submarine)
    padding_value = s.input_image[1, 1]
    width, height = size(s.input_image)

    padding_h = padding_value ? trues(1, height) : falses(1, height)
    padding_w = padding_value ? trues(width + 2, 1) : falses(width + 2, 1)

    s.input_image = [padding_h; s.input_image; padding_h]
    s.input_image = [padding_w s.input_image padding_w]
end

function interpret_num(s::Submarine, x::UInt, y::UInt)::UInt
    s.input_image[x - 1, y - 1] << 8 |
    s.input_image[x, y - 1] << 7 |
    s.input_image[x + 1, y - 1] << 6 |
    s.input_image[x - 1, y] << 5 |
    s.input_image[x, y] << 4 |
    s.input_image[x + 1, y] << 3 |
    s.input_image[x - 1, y + 1] << 2 |
    s.input_image[x, y + 1] << 1 |
    s.input_image[x + 1, y + 1]
end

function enhance!(s::Submarine)
    width, height = size(s.input_image)

    new_padding_value = s.input_image[1, 1] ? s.enhancement_algorithm[512] : s.enhancement_algorithm[1]

    result = new_padding_value ? trues(width, height) : falses(width, height)
    padding_changed = false

    for x::UInt in 2:width - 1
        for y::UInt in 2:height - 1
            num = interpret_num(s, x, y)
            result[x, y] = s.enhancement_algorithm[num + 1]

            if x == 2 || x == width - 1 || y == 2 || y == height - 1
                padding_changed = padding_changed || result[x, y] != new_padding_value
            end
        end
    end

    s.input_image = result

    padding_changed && extend_padding!(s)
end

function lit_pixels(s::Submarine)::UInt
    sum(s.input_image)
end

function print_image(s::Submarine)
    width, height = size(s.input_image)

    for x in 1:width
        for y in 1:height
            print(s.input_image[x, y] ? '#' : '.')
        end

        println()
    end
end

function output(s::Submarine)
    finish_padding!(s)

    for _ in 1:2
        enhance!(s)
    end

    println("Part 1: $(lit_pixels(s))")

    for _ in 1:48
        enhance!(s)
    end

    println("Part 2: $(lit_pixels(s))")
end

end